js/ui/appDisplay.js

```diff
@@ -647,19 +650,29 @@ const AllView = new Lang.Class({
                 this.addItem(icon);
         }));
 
-        // Add the App Center icon if it is enabled and installed
-        if (global.settings.get_boolean(EOS_ENABLE_APP_CENTER_KEY)) {
-            let app = appSys.lookup_app(EOS_APP_CENTER_ID);
-            if (app)
-                this.addItem(new AppCenterIcon(app));
-            else
-                log('App center ' + EOS_APP_CENTER_ID + ' is not installed');
-        }
-
+        // Add the App Center icon if it is enabled (and installed)
+        this._maybeAddAppCenterIcon();
 
         this.loadGrid();
     },
 
+    _maybeAddAppCenterIcon: function() {
+        if (this._appCenterIcon)
+            return;
+
+        if (!global.settings.get_boolean(EOS_ENABLE_APP_CENTER_KEY))
+            return;
+
+        let appSys = Shell.AppSystem.get_default();
+        if (!appSys.lookup_app(EOS_APP_CENTER_ID)) {
+            log('App center ' + EOS_APP_CENTER_ID + ' is not installed');
```

Comment:
I had the same thoughts, but turns out it can actually happen!: https://github.com/endlessm/gnome-shell/commit/1140d786a40085807da48e73f04714df2e83ca2b

Ask @wjt for details, but I'd leave it for now.